src/files/infra/clients/DirectUploadClient.ts

```diff
@@ -0,0 +1,220 @@
+import axios from 'axios'
+import { IDirectUploadClient } from '../../domain/clients/IDirectUploadClient'
+import { FileUploadDestination } from '../../domain/models/FileUploadDestination'
+import pLimit from 'p-limit'
+import {
+  buildRequestConfig,
+  buildRequestUrl
+} from '../../../core/infra/repositories/apiConfigBuilders'
+import * as crypto from 'crypto'
+import { IFilesRepository } from '../../domain/repositories/IFilesRepository'
+import { FileUploadError } from './errors/FileUploadError'
+import { FilePartUploadError } from './errors/FilePartUploadError'
+import { MultipartCompletionError } from './errors/MultipartCompletionError'
+import { AddUploadedFileToDatasetError } from './errors/AddUploadedFileToDatasetError'
+import { UrlGenerationError } from './errors/UrlGenerationError'
+import { MultipartAbortError } from './errors/MultipartAbortError'
+import { FileUploadCancelError } from './errors/FileUploadCancelError'
+import { ApiConstants } from '../../../core/infra/repositories/ApiConstants'
+
+export class DirectUploadClient implements IDirectUploadClient {
+  private filesRepository: IFilesRepository
+  private maxMultipartRetries: number
+
+  private readonly checksumAlgorithm: string = 'md5'
+
+  private readonly progressAfterUrlGeneration: number = 10
+  private readonly progressAfterFileUpload: number = 100
+
+  constructor(filesRepository: IFilesRepository, maxMultipartRetries = 5) {
+    this.filesRepository = filesRepository
+    this.maxMultipartRetries = maxMultipartRetries
+  }
+
+  public async uploadFile(
+    datasetId: number | string,
+    file: File,
+    progress: (now: number) => void,
+    abortController: AbortController,
+    destination?: FileUploadDestination
+  ): Promise<string> {
```

Comment:
What about returning an object like `Promise<{ storageId : string }>`, perhaps is more descriptive

Reply:
Personally, I don't like very much the idea of making the returned data structure more complex in favor of improving readability here, especially when we provide documentation to clearly understand which data is returned as a string.

Reply:
yep, sounds good, lets keep the string 👍🏼